server/app/survey_test.go

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package app
+
+import (
+	"testing"
+	"time"
+
+	mmModel "github.com/mattermost/mattermost/server/public/model"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+
+	"github.com/mattermost/mattermost-plugin-user-survey/server/model"
+)
+
+func TestSaveSurvey(t *testing.T) {
+	t.Run("base case - empty survey should fail", func(t *testing.T) {
+		th := SetupAppTest(t)
+
+		err := th.App.SaveSurvey(&model.Survey{})
+		require.Error(t, err, "should return error as empty survey is invalid")
+	})
+
+	t.Run("should successfully save a valid survey", func(t *testing.T) {
+		th := SetupAppTest(t)
+		th.MockedStore.On("SaveSurvey", mock.Anything).Return(nil)
```

Comment:
I'm kind of surprised that we mock the store here given that that's part of the plugin. I know you'd need to do that for causing error cases, but I wouldn't think you'd need it for success cases.

Is that just for these tests since it involves error cases, or will you always need to mock the store?

Reply:
Generally I just mock store layer in app layer's unit tests. Store later is supposed to have its own unit tests against read database, but at the moment they don't exist (the framework for adding them does, though). So the idea is at some point I add store layer tests as well, and combined, the two unit tests would be complete. Does that resolve your concern?

Reply:
Personally, I tend to prefer not using mocks where possible because it covers more of the code at once and it does it in more of a realistic way. I ran into that with the MS Teams plugin testing where, with mocks, it felt like I was basically testing things like "calling `saveSomething` should call `store.SaveSomething`" which confirms that the code doesn't contain typos, but not that it did what I want it to. With a real store, I could change that to "calling `saveSomething` sets a value in the DB and returns another value" because I'm not having to pretend how `store.SaveSomething` behaves. Hopefully that example makes some sense and isn't too hand wavey 😅

Also, I tend to find mocking is a bit of a pain to do which is another reason I avoid it when possible

Reply:
Consider these as unit tests. I can add the integration tests later as per time. Created https://mattermost.atlassian.net/browse/MM-58103 to track this.

Reply:
That works, and sorry, I didn't mean to sound like we needed to change all these now. I've just recently liked talking about tests and how I think they can be done better

Reply:
No worries, I got your point.